target/linux/ipq40xx/base-files/lib/preinit/05_set_iface_mac_ipq40xx.sh

```diff
@@ -26,6 +26,11 @@ preinit_set_mac_address() {
 		mac_lan=$(get_mac_binary "/sys/bus/i2c/devices/0-0050/eeprom" 0x66)
 		[ -n "$mac_lan" ] && ip link set dev eth0 address "$mac_lan"
 		;;
+	mikrotik,wap-ac)
+		base_mac=$(cat /sys/firmware/mikrotik/hard_config/mac_base)
+		ip link set dev eth0 address "$base_mac"
+		ip link set dev eth1 address $(macaddr_add "$base_mac" 1)
```

Comment:
This could probably be moved to NVMEM as to me it looks like the base MAC is always at the same offset in hard_config

Reply:
I’d prefer to go through the hard_config abstraction. That’s what it’s for, after all. Kept as-is.

Reply:
Well, we should get rid of scripts if there is a standardized kernel interface not expand them further, that's my standing on any script.

Would love opinions from others as well

Reply:
I don’t think that nvmem can be that standardized kernel interface. hard_config is structured as (magic, n * element = (tag, length, value)). Sure, on all of my devices, I find the base MAC address at hard_config + 0x10 (tag at +0xc), but that’s not even the first element, it’s preceded by something unknown with tag 0x1a.

```
00000000  48 61 72 64 1a 00 04 00  00 00 00 00 04 00 08 00  |Hard............|
00000010  08 55 31 xx yy zz 00 00  0e 00 04 00 04 00 00 00  |.U1.............|
```

Directing nvmem to hard_config + 0x10 means that we wouldn’t be properly interpreting this structure at all. Nobody’s promised us that elements will appear in any particular order, or that other elements won’t precede the base MAC address—in fact, there’s something else before it here. Nobody’s promised us that whatever precedes the base MAC address will forever have the same length. And with all of this, I’m mindful that we’ve seen other cases where MikroTik’s made config changes during a production run without changing the model number or revision, as with the [BDF fiasco](https://github.com/openwrt/openwrt/pull/4679). They could easily make a change during the production run such that the MAC base winds up at some other offset in hard_config, and because it’s “just” a MAC address, it wouldn’t even necessarily be immediately apparent to anyone that OpenWrt was reading out garbage instead of a genuine MAC address.

So while I agree that a supported kernel interface would beat handling this in scripts (whether it’s `/lib/preinit/05_set_iface_mac_ipq40xx.sh` or `/etc/board.d/02_network`), I feel pretty strongly that there isn’t a suitable kernel interface for this outside of something that properly interprets hard_config, and I don’t think that’s within nvmem’s ability.

Reply:
@robimarko the assumption that mac is always at the same place in hard_config is broken by design. As you probably remember, the same assumption was made in ath79, only to be eventually crushed by reality, leading to the development of the rb_hardconfig driver.
I'm not sure what makes you think it would be a good idea to go back to this assumption which is bound to be disproven eventually. As you know, the `hard_config` block data is structured in "tags" and the structure imparts no requirement on a specific tag position. I think it would be completely backwards to ignore the internal structure *again*. My 2c.

Reply:
I understand, it's MikroTik being special over and over.
Fine, keep the script, though I seriously think that it would be great for the MikroTik parser to also provide NVMEM cells so that kernel can fetch the MAC directly.

Its the same issue as using ONIE TLV format has:
https://lore.kernel.org/lkml/20200831172443.GD2568@plvision.eu/t/

Reply:
Interesting, I wasn't aware of this possibility. I can certainly look into implementing this into ˋrb_hardconfig` down the road.

Reply:
It would be ideal, cause then NVMEM cells could be used to fetch everything in the kernel space instead of having to go through userspace

Reply:
The one "minor" caveat I see is that it seems that none of the proposed changes in the patch you quoted have been accepted (nor its [parser variant sequel](https://lore.kernel.org/lkml/20210608190327.22071-4-vadym.kochan@plvision.eu/)), which doesn't exactly bode well for another implementation that would basically have the same constraints.
Until then, it's probably best to see what canonical approach is accepted (if any)? Meanwhile, the scripts work, even if they aren't perfect.

Reply:
ok. Feel free to ping me again when it's gotten through, then I'll look into implementing for our use case :)

Reply:
Ok so for now preinit script is the only way. Maybe add a comment so this ""workaround"" doesn't became a permanent solution? Just to have a reference that it should be fixed with a proper and better solution

Reply:
it's currently the *only* solution. If/when a better alternative appears, it will be applied to all mikrotik devices anyway :)